modules/apps/commerce/commerce-product-service/src/main/java/com/liferay/commerce/product/internal/security/permission/resource/CommerceChannelRelModelResourcePermission.java

```diff
@@ -0,0 +1,161 @@
+
+import com.liferay.commerce.discount.model.CommerceDiscount;
+import com.liferay.commerce.discount.permission.CommerceDiscountPermission;
+import com.liferay.commerce.discount.service.CommerceDiscountLocalService;
+import com.liferay.commerce.product.model.CommerceChannelRel;
+import com.liferay.commerce.product.service.CommerceChannelRelLocalService;
+import com.liferay.portal.kernel.exception.PortalException;
+import com.liferay.portal.kernel.security.permission.PermissionChecker;
+import com.liferay.portal.kernel.security.permission.resource.ModelResourcePermission;
+import com.liferay.portal.kernel.security.permission.resource.PortletResourcePermission;
+import com.liferay.portal.kernel.service.ClassNameLocalService;
+
+import java.util.Objects;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * @author Luca Pellizzon
+ */
+@Component(
+	enabled = false, immediate = true,
+	property = "model.class.name=com.liferay.commerce.product.model.CommerceChannelRel",
+	service = ModelResourcePermission.class
+)
+public class CommerceChannelRelModelResourcePermission
+	implements ModelResourcePermission<CommerceChannelRel> {
+
+	@Override
+	public void check(
+			PermissionChecker permissionChecker,
+			CommerceChannelRel commerceChannelRel, String actionId)
+		throws PortalException {
+
+		if (Objects.equals(
+				commerceChannelRel.getClassNameId(),
+				commerceDiscountClassNameId)) {
+
+			CommerceDiscount commerceDiscount =
+				commerceDiscountLocalService.getCommerceDiscount(
+					commerceChannelRel.getClassPK());
+
+			commerceDiscountPermission.check(
+				permissionChecker, commerceDiscount.getCommerceDiscountId(),
+				actionId);
+		}
+	}
+
+	@Override
+	public void check(
+			PermissionChecker permissionChecker, long commerceChannelRelId,
+			String actionId)
+		throws PortalException {
+
+		CommerceChannelRel commerceChannelRel =
+			commerceChannelRelLocalService.getCommerceChannelRel(
+				commerceChannelRelId);
+
+		if (Objects.equals(
+				commerceChannelRel.getClassNameId(),
+				commerceDiscountClassNameId)) {
+
+			CommerceDiscount commerceDiscount =
+				commerceDiscountLocalService.getCommerceDiscount(
+					commerceChannelRel.getClassPK());
+
+			commerceDiscountPermission.check(
+				permissionChecker, commerceDiscount.getCommerceDiscountId(),
+				actionId);
+		}
+	}
+
+	@Override
+	public boolean contains(
+			PermissionChecker permissionChecker,
+			CommerceChannelRel commerceChannelRel, String actionId)
+		throws PortalException {
+
+		if (Objects.equals(
+				commerceChannelRel.getClassNameId(),
+				commerceDiscountClassNameId)) {
+
+			CommerceDiscount commerceDiscount =
+				commerceDiscountLocalService.getCommerceDiscount(
+					commerceChannelRel.getClassPK());
+
+			return commerceDiscountPermission.contains(
+				permissionChecker, commerceDiscount.getCommerceDiscountId(),
+				actionId);
+		}
+
+		return false;
```

Comment:
@dsitu now that the model resource permission is using the generic model, we are left with one big question...
Should we return false or true if we get to this point and the class is not a managed one?
Right now we are managing only discounts, but inside the ChannelRel we can have anything.
We think this method should not be invoked with unmanaged classes, but still, you never know.
What do you think about it?
cc @riccardo-alberti

Reply:
I think returning false is the way to go. That way if anyone tries to use this for an unmanaged one, they would notice that it fails and will have to implement the logic.